### PR TITLE
[Crop and Soil] Adding e2e for C&S module

### LINKED ISSUE #1993: [E2E Testing][C&S] Add end-to-end testing coverage for the Crop and Soil module
## Status

The end-to-end testing routine for RuFaS currently does not cover the Crop and Soil module

## Intention

End-to-end testing should cover the Crop and Soil module in order for maintainers to more easily verify that correct behavior is maintained in when reviewing changes.

## Request

Add the following files to the directory `input/data/end_to_end_testing`:
- `e2e_json_crop_soil_filter.json`. This file should contain a filter for gathering actual results from the crop and soil module, as well as contain the expected C&S-related results. Use `e2e_json_feed_storage_filter.json` as an example for how to write this file.
- `e2e_comparison_crop_soil_differences.json`. This file should contain a filter that will gather the results of end-to-end testing the Crop and Soil module. Use `e2e_comparison_feed_storage_differences.json` as an example for how to write this file.
- `end_to_end_testing_result_paths.json`. This file already exists, but an entry should be added that specifies the Crop and Soil domain, the path to expected outputs, and the pattern used to find the actual outputs from the C&S module. Use the existing entry for Feed Storage as an example for how to write this.

### PR DESCRIPTION
**Pull request by matthew7838**
Adds crop and soil module (domain) end-to-end testing expected results, associated filter files, and updated testing metadata expected.
### Context

Issue(s) closed by this pull request: closes #1993 

### What

Adds testing results, filters, and configs for CS module end-2-end testing.

### Why

Adds testing results, filters, and configs for crop and soil module end-2-end testing.

### How

1.  Set up the initial filters and expected results for crop and soil domain in `e2e_json_crop_soil_filter.json`
2. The start and end dates for the e2e testing config had to be updated to 2013-2014 because the previous dates were causing net merit errors.
3. e2e testing expected results for crop and soil were updated based on results of the new date range for the e2e config.
4. Add all C&S required metadata to the e2e testing metadata in order to enable simulation run.

### Test plan

e2e testing was run as outlined in the issue and there were no found differences between the expected and the actual results for crop and soil domain.

### CODE REVIEW

**Comment by ew3361zh** [comment]
I haven't tested it yet but I think it's worth minifying the expected results so we're not adding 550K+ lines.

For Animal and FeedStorage expected results, I turned on `minify_output_file` in OM.dict_to_file_json() - which is the method by which the expected results are saved - and then ran end-2-end testing. 

You can then go to the results in the JSONs folder and copy-paste them into the expected results section of your filter file.

Screenshot 2024-11-06 at 8 50 45 AM

File: input/metadata/end_to_end_testing_metadata.json [hunk 1]
```diff
@@ -153,6 +153,62 @@
...
+      "type": "json",
+      "properties": "soil_profile_properties"
+    },
+    "soil_2": {
+      "title": "Soil data",
+      "description": "Characteristic soil information, including composition, slope, and layer details.",
+      "path": "input/data/soil/default_soil_2.json",
+      "type": "json",
+      "properties": "soil_profile_properties"
+    },
```
Comment (Isgotkowitz) [inline input/metadata/end_to_end_testing_metadata.json]:
```suggestion
```
This second soil does not need to be included.

File: input/metadata/end_to_end_testing_metadata.json [hunk 2]
```diff
@@ -153,6 +153,62 @@
...
+      "type": "csv",
+      "properties": "animal_net_merit_properties"
+    },
+    "animal_top_listing_semen": {
+      "title": "Animal Top Listing Semen data",
+      "description": "The top listing semen value for new born calves.",
+      "path": "input/data/animal/animal_genetics/TopListingSemen_HO.csv",
+      "type": "csv",
+      "properties": "animal_top_listing_semen_properties"
+    },
```
Comment (Isgotkowitz) [inline input/metadata/end_to_end_testing_metadata.json]:
```suggestion
```
These are duplicates

File: changelog.md [hunk 3]
```diff
@@ -39,6 +39,7 @@ v0.9.2
 - [2033](https://github.com/RuminantFarmSystems/MASM/pull/2033) - [minor change] [InputManager] Consolidate methods for adding data to the Input Manager at runtime.
 - [2039](https://github.com/RuminantFarmSystems/MASM/pull/2039) - [minor change] [Animal][Testing] Renames AnimalNutrients to Nutrients and adds full unit testing coverage for this class.
 - [2014](https://github.com/RuminantFarmSystems/MASM/pull/2014) - [minor change] [Feed Storage] Fixes the Feed Storage mass degradation routines to properly account for mass and moisture loss, and updates E2E test suite to test these changes.
+- [2054](https://github.com/RuminantFarmSystems/MASM/pull/2054) - [minor change] [e2e Testing] Adding e2e testing covergae for C&S module.
```
Comment (Isgotkowitz) [inline changelog.md]:
```suggestion
- [2054](https://github.com/RuminantFarmSystems/MASM/pull/2054) - [minor change] [e2e Testing] Adding end-to-end testing coverage for the Crop and Soil module.
```

**Review by Isgotkowitz — COMMENTED** [review]
For whatever reason the Animal and C&S E2E tests are not passing. I skimmed the differences and it seemed to just be floating point stuff, so I think it would be ok to update the expected results using whatever is in the actual outputs. Once those two areas are passing, I think it will be good to go!

**Comment by matthew7838** [comment]
> For whatever reason the Animal and C&S E2E tests are not passing. I skimmed the differences and it seemed to just be floating point stuff, so I think it would be ok to update the expected results using whatever is in the actual outputs. Once those two areas are passing, I think it will be good to go!

That's is weird, I'm seeing C&S working fine on my side.

**Review by Isgotkowitz — APPROVED** [review]
For me, the Animal and Crop & Soil tests do not pass right out of the box. However, this is only because of some floating point differences, and both domains pass if I change the significant digits argument to 6 in the `DeepDiff` init in `E2ETestResultsComparer`. This sees like a reasonable level of precision to me, so looks good! Nice job Matthew.

Note that the significant digits argument does not need to be changed in this PR, and I think we should make determining that value its own PR/process

**Comment by ew3361zh** [comment]
Neither Animal nor Crop and Soil testing were passing for me. The majority of issues were floating point differences but they did not clear up for me when I reset the significant_digits to 6.

Also, in both results JSONs, there were issues with items added/removed. I feel like those should be addressed before merging. My suggestion would be rerunning the e2e testing and readding the results to their respective expected sections.

Strangely, they were different items each time too (and different amounts of items) with runs done a couple minutes from one another. Going to reprint below for reference:

First run Animal:
Screenshot 2024-11-07 at 3 25 07 PM

Second run Animal:
Screenshot 2024-11-07 at 3 27 57 PM

First run C&S:
Screenshot 2024-11-07 at 3 25 16 PM

Second run C&S:
Screenshot 2024-11-07 at 3 28 10 PM

It's really strange that there are different results for each run. I can help look into this if you're not able to figure out what's going on.

**Comment by ew3361zh** [comment]
Also curious what was changed (and why) in the `e2e_json_animal_filter.json` file for this PR.

**Comment by matthew7838** [comment]
> Also curious what was changed (and why) in the `e2e_json_animal_filter.json` file for this PR.

 I tried to change the expected result on my end to match.

**Comment by matthew7838** [comment]
> Neither Animal nor Crop and Soil testing were passing for me. The majority of issues were floating point differences but they did not clear up for me when I reset the significant_digits to 6.
> 
> Also, in both results JSONs, there were issues with items added/removed. I feel like those should be addressed before merging. My suggestion would be rerunning the e2e testing and readding the results to their respective expected sections.
> 
> Strangely, they were different items each time too (and different amounts of items) with runs done a couple minutes from one another. Going to reprint below for reference:
> 
> First run Animal: Screenshot 2024-11-07 at 3 25 07 PM
> 
> Second run Animal: Screenshot 2024-11-07 at 3 27 57 PM
> 
> First run C&S: Screenshot 2024-11-07 at 3 25 16 PM
> 
> Second run C&S: Screenshot 2024-11-07 at 3 28 10 PM
> 
> It's really strange that there are different results for each run. I can help look into this if you're not able to figure out what's going on.

Yea I am not sure about this, one thing I would check is if the second and third run are the same.

**Comment by Isgotkowitz** [comment]
> Neither Animal nor Crop and Soil testing were passing for me. The majority of issues were floating point differences but they did not clear up for me when I reset the significant_digits to 6.
> 
> Also, in both results JSONs, there were issues with items added/removed. I feel like those should be addressed before merging. My suggestion would be rerunning the e2e testing and readding the results to their respective expected sections.
> 
> Strangely, they were different items each time too (and different amounts of items) with runs done a couple minutes from one another. Going to reprint below for reference:
> 
> First run Animal: Screenshot 2024-11-07 at 3 25 07 PM
> 
> Second run Animal: Screenshot 2024-11-07 at 3 27 57 PM
> 
> First run C&S: Screenshot 2024-11-07 at 3 25 16 PM
> 
> Second run C&S: Screenshot 2024-11-07 at 3 28 10 PM
> 
> It's really strange that there are different results for each run. I can help look into this if you're not able to figure out what's going on.

Really, really strange. I reran the E2E tests on both Linux and Windows, neither exhibited this behavior.

**Review by ew3361zh — APPROVED** [review]
I think this looks good. I was not able to resolve it failing on my side but we had @allisterakun test this on his Mac and he was getting successful results so I think this may be an issue just for my setup. I will approve.

Please do write up an issue to address the significant digits/tolerance problems.

**Pull request merged**